server/src/main/java/org/elasticsearch/action/search/AbstractSearchAsyncAction.java

```diff
@@ -801,48 +800,50 @@ protected final ShardSearchRequest buildShardSearchRequest(SearchShardIterator s
         return shardRequest;
     }
 
+    protected record ShardRequest(int shardIndex, SearchShardIterator shardIt, SearchShardTarget shard) {}
+
     /**
      * Returns the next phase based on the results of the initial search phase
      */
     protected abstract SearchPhase getNextPhase();
 
-    private static final class PendingExecutions {
+    private final class PendingExecutions {
```

Comment:
what requires making this non static? The call to `doPerformPhaseOnShard`? Isn't it easier to follow as a static class instead of tying it to a specific instance of the action?

Reply:
When we coded this it seemed cleaner to just tie it to the specific class and avoid a layer of indirection by passing the class in via some listener/releasable/you name it.
Either way, both the current code and this are needless complex. We should in a cleanup step flatten the map of these things into the `AbstractSearchAsyncAction` I'd say and avoid all the circular references here to begin with.
Maybe just go for this as it's the shortest for now and we promise we'll clean it up :)?

Reply:
I don't know, it does not seem like a great cleanup to me. But it is important to get away from the consumer? I am not sure I follow.

Reply:
IMO It's needed because once we have a Consumer we can only `accept` and therefore call the `doPerfomPhaseOnShard`, while with `ShardRequest` (naming TBD) we can choose later on what to do with it.
One example: if we would like to batch requests but we have just one request for a specific data node, we cannot know that in advance (without looping `shardsIts` a second time) and for the single request we should be using the non-bulk path.

Also batching becomes awkward. You'd need to create another "pending execution queue" to collect all the task on a per data node basis.

Reply:
@javanna yea this is not about cleanup (though admittedly my choice of `Consumer<Releasable>` has always been weird :P), this is about being able to merge these instances.